node/subsystem-util/src/validator_discovery.rs

```diff
@@ -104,6 +105,79 @@ async fn connect_to_authorities<Context: SubsystemContext>(
 	Ok((connected_rx, revoke_tx))
 }
 
+/// A struct that assists performing multiple concurrent connection requests.
+///
+/// This allows concurrent connections to validator sets at different `relay_parents`
+/// and multiplexes their results into a single `Stream`.
+#[derive(Default)]
+pub struct ConnectionRequests {
+	// added connection requests relay_parent -> StreamUnordered token
+	id_map: HashMap<Hash, usize>,
+
+	// Connection requests themselves.
+	requests: StreamUnordered<ConnectionRequest>,
+}
+
+impl ConnectionRequests {
+	/// Insert a new connection request.
+	///
+	/// If a `ConnectionRequest` under a given `relay_parent` already exists it will
+	/// be revoked and substituted with a new one.
+	pub fn put(&mut self, relay_parent: Hash, request: ConnectionRequest) {
+		self.remove(&relay_parent);
+		let token = self.requests.push(request);
+
+		self.id_map.insert(relay_parent, token);
+	}
+
+	/// Remove a connection request by a given `relay_parent`.
+	pub fn remove(&mut self, relay_parent: &Hash) {
+		if let Some(token) = self.id_map.remove(relay_parent) {
+			Pin::new(&mut self.requests).remove(token);
+		}
+	}
+}
+
+impl stream::Stream for ConnectionRequests {
+	/// (relay_parent, validator_id, peer_id).
+	type Item = (Hash, ValidatorId, PeerId);
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut task::Context) -> Poll<Option<Self::Item>> {
+		// If there are currently no requests going on, pend instead of
+		// polling `StreamUnordered` which would lead to it terminating
+		// and returning `Poll::Ready(None)`.
+		if self.requests.is_empty() {
+			return Poll::Pending;
+		}
+
+		match Pin::new(&mut self.requests).poll_next(cx) {
+			Poll::Ready(Some((yielded, token))) => {
+				match yielded {
+					StreamYield::Item(item) => {
+						if let Some(relay_parent) = self.id_map.iter()
+							.find_map(|(relay_parent, &val)|
+								if val == token {
+									Some(relay_parent)
+								} else {
+									None
+								}
+							) {
+							return Poll::Ready(Some((*relay_parent, item.0, item.1)));
+						}
+					}
+					StreamYield::Finished(_) => {
+						// Unreachable: `ConnectionRequest` never finishes
```

Comment:
Is this true? As far as I remember, the request returns `None` when all requests are fulfilled.

Reply:
https://github.com/paritytech/polkadot/blob/fe6ec9a9e294e0cf567fb482a5ca8d18dd86186c/node/subsystem-util/src/validator_discovery.rs#L200

Reply:
Here we end the connection request

Reply:
Ah, right, I've must have checked an old version. It's not unreachable, but we still want to return `Pending` here and not drop the connection request.

Reply:
Yeah